src/services/factories/tokenList.ts

```diff
@@ -88,32 +99,124 @@ export function getTokensFactory(factoryParams: {
   }
 
-  async function updateTokenIds(networkId: number): Promise<void> {
-    // Set as updated on start to prevent concurrent queries
-    updatedIdsForNetwork.add(networkId)
+  async function getErc20DetailsOrAddress(networkId: number, tokenAddress: string): Promise<TokenFromErc20 | string> {
+    // Simple wrapper function to return original address instead of null for make logging easier
+    const erc20Details = await getTokenFromErc20({ networkId, tokenAddress })
+    return erc20Details || tokenAddress
+  }
+
+  async function fetchAddressesAndIds(networkId: number, numTokens: number): Promise<Map<string, number>> {
+    logDebug(`[tokenListFactory][${networkId}] Fetching addresses for ids from 0 to ${numTokens - 1}`)
+
+    const promises = Array.from({ length: numTokens }, (_, tokenId) =>
+      exchangeApi.getTokenAddressById({ networkId, tokenId }),
+    )
+
+    const tokenAddresses = await Promise.all(promises)
+
+    return new Map(tokenAddresses.map((tokenAddress, id) => [tokenAddress, id]))
+  }
+
+  async function updateTokenDetails(networkId: number, numTokens: number): Promise<void> {
+    // Fetch addresses from contract given numTokens count
+    const addressesAndIds = await retry({
+      fn: fetchAddressesAndIds,
+      fnParams: [networkId, numTokens],
+    })
+    logDebug(`[tokenListFactory][${networkId}] Token id and address mapping:`)
+    addressesAndIds.forEach((id, address) => logDebug(`[tokenListFactory][${networkId}] ${id} : ${address}`))
+
+    // Create a map of current token list addresses and tokens
+    const localAddressesMap = new Map<string, TokenDetails>(
+      tokenListApi
+        // Get a copy of the current token list
+        .getTokens(networkId)
+        // Remove tokens that are on the list but not registered on the contract
+        .filter(({ address }) => addressesAndIds.has(address))
+        // Map it with token address as key for easy access
+        .map(token => [token.address, token]),
+    )
+
+    const promises: Promise<TokenFromErc20 | string>[] = []
+
+    // Go over all value (id) and key (tokenAddress) pairs registered on the contract
+    addressesAndIds.forEach((id, tokenAddress) => {
+      if (!localAddressesMap.has(tokenAddress)) {
+        // New token not in our local list, fetch erc20 details for it
+        logDebug(`[tokenListFactory][${networkId}] Address ${tokenAddress} with id ${id} not in local list, fetching`)
+        promises.push(getErc20DetailsOrAddress(networkId, tokenAddress))
+      } else {
+        // Token already exists, update id
+        logDebug(`[tokenListFactory][${networkId}] Address ${tokenAddress} already in the list, updating id to ${id}`)
+        const token = localAddressesMap.get(tokenAddress) as TokenDetails
+        token.id = id
+      }
+    })
+
+    const partialTokens = await Promise.all(promises)
+
+    // For all newly fetched tokens
+    partialTokens.forEach(partialToken => {
+      if (typeof partialToken === 'string') {
```

Comment:
Can `partialToken` ever be `null`? or anything `falsy`? If not, can just do `else` after first `if`, no need for `partialToken` check

Reply:
It cannot be null because on the wrapper promise above I replace `null` by the `tokenAddress`.
But, the type of the response can still be `null`, that's why I have the additional check after this.

I'm happy yo update that if I manage to make TS understand `partialToken` cannot be null.